pandas/core/indexes/datetimes.py

```diff
@@ -1138,7 +1153,7 @@ def _maybe_utc_convert(self, other):
                 raise TypeError('Cannot join tz-naive with tz-aware '
                                 'DatetimeIndex')
 
-            if self.tz != other.tz:
+            if not self._tz_compare(other):
```

Comment:
can you also add for *all* the other cases of ``str(self.tz) != str(other)`` (or variants on this)

Reply:
The other places where pandas uses variants for `str(self.tz) != str(other.tz)` isn't particularly for DatetimeIndex. For example in `test_timezones.py` this is used to test the equality of `Timestamp` I guess we have to write a more general method which takes in any object with a .tz attribute and compare it. In that case which file should I put this method in?

Reply:
the right place to put this is in
pandas/_libs/tslibs/timezones.pyx

and just call it from where needed

Reply:
FYI - I added an additional instance of this in a commit that went in earlier today.  Current location on `master`:

https://github.com/pandas-dev/pandas/blob/288bf6e5e3287a8c66a7ed41050db0ebf56d7ea6/pandas/core/indexes/interval.py#L242-L243